spouse/spouse_demo.py

```diff
@@ -0,0 +1,363 @@
+#
+# Now, we'll train a model of the LFs to estimate their weights and combine their outputs. Once the model is trained, we can combine the outputs of the LFs into a single, noise-aware training label set for our extractor.
+
+# %%
+from snorkel.labeling.model.label_model import LabelModel
+
+label_model = LabelModel(cardinality=2, verbose=True)
+label_model.fit(train_L, Y_dev, n_epochs=5000, log_freq=500, seed=12345)
+
+# %% [markdown]
+# ### Label Model Metrics
+# Since our dataset is highly unbalanced (91% of the labels are negative), even a trivial baseline that always outputs negative can get a high accuracy. So we evaluate the label model using the F1 score and ROC-AUC rather than accuracy.
+
+# %%
+from snorkel.analysis.metrics import metric_score
+from snorkel.analysis.utils import probs_to_preds
+
+Y_probs_dev = label_model.predict_proba(dev_L)
+Y_preds_dev = probs_to_preds(Y_probs_dev)
+print(
+    f"Label model f1 score: {metric_score(Y_dev, Y_preds_dev, probs=Y_probs_dev, metric='f1')}"
+)
+print(
+    f"Label model roc-auc: {metric_score(Y_dev, Y_preds_dev, probs=Y_probs_dev, metric='roc_auc')}"
+)
+
+# %%
+import matplotlib.pyplot as plt
+
+Y_probs_train = label_model.predict_proba(train_L)
+plt.hist(Y_probs_train[:, 1], bins=20, range=(0.0, 1.0))
+plt.show()
+
+# %% [markdown]
+# ### Part 4: Training our End Extraction Model
+#
+# In this final section of the tutorial, we'll use our noisy training labels alongside the development set labels to train our end machine learning model. We start by filtering out training examples which did not recieve a label from any LF, as these examples contain no signal. Then we concatenate them with dev set examples.
+#
+# %%
+from snorkel.analysis.utils import preds_to_probs
+from snorkel.labeling.utils import filter_unlabeled_dataframe
+
+# Change dev labels 1D array to 2D probabilities array as required for training end model.
+Y_probs_dev = preds_to_probs(Y_dev, 2)
+
+Y_probs_train = label_model.predict_proba(train_L)
+df_train_filtered, Y_probs_train_filtered = filter_unlabeled_dataframe(
+    X=df_train, y=Y_probs_train, L=train_L
+)
+
+df_combined = pd.concat([df_dev, df_train_filtered])
+Y_probs_combined = np.concatenate([Y_probs_dev, Y_probs_train_filtered], 0)
+
+# %% [markdown]
+# ### II. Training a _Long Short-term Memory_ (LSTM) Neural Network
+#
+# [LSTMs](https://en.wikipedia.org/wiki/Long_short-term_memory) can acheive state-of-the-art performance on many text classification tasks. We'll train a simple LSTM model below. tf_model contains functions for processing features and building the tensorflow graphs for training and evaluation.
+#
+# For this tutorial, we will be training a fairly effective deep learning model. More generally, however, Snorkel plugs in with many ML libraries, making it easy to use almost any state-of-the-art model as the end model!
+
+# %%
+from tf_model import get_model
+
+
+# Truncate sentences to limit memory usage when padding.
```

Comment:
add a a comment about what this feature means?

Reply:
Moved to tf_model, and added types and docstring.